dpx/src/dpx_pdfobj.rs

```diff
@@ -1110,159 +1152,125 @@ impl Drop for pdf_array {
-    let mut data = dict.data as *mut pdf_dict;
-    while !(*data).key.is_null() {
-        if key.as_ref() == pdf_name_value(&*(*data).key).to_bytes() {
-            /* Release the old value */
-            pdf_release_obj((*data).value);
-            (*data).value = value;
-            return 1i32;
-        }
-        data = (*data).next
+    if let Some(existing) = dict.inner.insert(PdfName::new(key), value) {
+        pdf_release_obj(existing);
+        1
+    } else {
+        0
     }
-    let name = pdf_new_name(key);
-    /*
-     * We didn't find the key. We build a new "end" node and add
-     * the new key just before the end
-     */
-    let new_node =
-        new((1_u64).wrapping_mul(::std::mem::size_of::<pdf_dict>() as u64) as u32) as *mut pdf_dict;
-    (*new_node).key = 0 as *mut pdf_obj;
-    (*new_node).value = 0 as *mut pdf_obj;
-    (*new_node).next = 0 as *mut pdf_dict;
-    (*data).next = new_node;
-    (*data).key = name;
-    (*data).value = value;
-    0i32
 }
+
 /* pdf_merge_dict makes a link for each item in dict2 before stealing it */
 #[no_mangle]
 pub unsafe extern "C" fn pdf_merge_dict(dict1: &mut pdf_obj, dict2: &pdf_obj) {
-    assert!(dict1.is_dict());
-    assert!(dict2.is_dict());
-    let mut data = dict2.data as *const pdf_dict;
-    while !(*data).key.is_null() {
-        pdf_add_dict(
-            dict1,
-            //pdf_link_obj((*data).key),
-            pdf_name_value(&*(*data).key).to_bytes(),
-            pdf_link_obj((*data).value),
-        );
-        data = (*data).next
+    let dict1 = dict1.get_dict_mut();
+    let dict2 = dict2.get_dict();
+    for (k, &v) in dict2.inner.iter() {
+        let linked = pdf_link_obj(v);
+        dict1.inner.entry(k.clone())
+            .and_modify(|existing| {
+                let existing = mem::replace(existing, linked);
+                pdf_release_obj(existing);
+            })
+            .or_insert(linked);
     }
 }
+
 #[no_mangle]
 pub unsafe extern "C" fn pdf_foreach_dict(
     dict: &mut pdf_obj,
     mut proc_0: Option<
-        unsafe extern "C" fn(_: *mut pdf_obj, _: *mut pdf_obj, _: *mut libc::c_void) -> i32,
+        unsafe extern "C" fn(_: PdfObjRef, _: PdfObjRef, _: *mut libc::c_void) -> i32,
     >,
     mut pdata: *mut libc::c_void,
 ) -> i32 {
-    let mut error: i32 = 0i32;
-    assert!(proc_0.is_some());
-    assert!(dict.is_dict());
-    let mut data = dict.data as *mut pdf_dict;
-    while error == 0 && !(*data).key.is_null() {
-        error = proc_0.expect("non-null function pointer")((*data).key, (*data).value, pdata);
-        data = (*data).next
-    }
-    error
+    let proc = proc_0.expect("non-null function pointer");
+    let dict = dict.get_dict_mut();
+    let mut make_key_ptr = |key: &PdfName| safe_new_obj(PdfObjType::NAME, |object| {
+        object.data = Box::into_raw(Box::new(key.clone())) as *mut libc::c_void;
+    });
+    dict.foreach_dict(|k, v, pdata| proc(make_key_ptr(k), v, pdata), pdata)
```

Comment:
This is the only thing I can find that's really different --

* The original calls proc with the actual PdfObjRef (which happens to be a PdfName)
* This code makes a new PdfObjRef to call proc with, and also happens to never drop it.